antsibull/schemas/docs/base.py

```diff
@@ -441,6 +453,7 @@ class DocSchema(BaseModel):
     todo: t.List[str] = []
     version_added: str = 'historical'
     version_added_collection: str = COLLECTION_NAME_F
+    attributes: t.Dict[str, AttributeSchema] = {}
```

Comment:
Okay, I think I see why you want a TypedDict.  It's for this structure, right?  My mind isn't 100% in the pydantic headspace right now but I think something like this might work?

``` python
class AttributeSchema(BaseModel):
   [..]

class ActionGroupSchema(AttributeSchema):
    membership: t.List[str] = []
[..] # Other Attributes that require something extra


ActionAttributeSchema = AttributeSchema
[..] # Other Attributes that only have the base info


class AttributeContainerSchema(BaseModel):
    action: t.Optional[ActionAttributeSchema] = None
    action_group: t.Optional[ActionGroupAttributeSchema] = None
    [..]

class DocSchema(BaseModel):
    [..]
    attributes: AttributeContainerSchema = AttributeContainerSchema({})
```

Reply:
This will partially work, at least I would hope so. One downside is that for entries which don't use any of the extra values, a random one of these might be used, since they all match (since the additional values have default values).

Reply:
Ah sorry, I didn't look too closely at the last two things, which were the important ones :) Yes, this would solve the problem. **But** it requires that we know ahead all possible attributes that will be ever defined. Which unfortunately isn't how this system works, since everyone can define their own attributes.

Reply:
Hmm.... I can imagine what a solution would look like but it doesn't feel good in my gut.  Have to analyze (1) if the things that feel bad are the only logical outcomes of having arbitrary fieldnames in this part of the schema and (2) if it's possible to implement what the solution looks like.

So, let's say there's a schema side (validating and normalizing the data) and a programming side (using the data after it's been normalized by the schema) to this problem.  On the programming side, if AttributeContainerSchema is analagous to a defaultdict (either a defaultdict or a class whose attribute lookup has similar rules to a defaultdict) then we could have unknown keys return Attribute instances with all default values.  Code which deals with attributes does something like this pseudo code:
``` python
for attribute in plugin_data['attributes']:
    if attribute['supports'] != 'none':
        # do thing with attribute
```

On the schema side... Arbitrary field names are going to mean we have to set `extra=Extra.allow` on the AttributeContainerSchema.  We still want to validate and normalize that these extra fields have all the fields in AttributeSchema.  The schema for that is probably a separate class that is the same as AttributeSchema but  has `extra=Extra.allow`.  I can't think of a way to associate those to the extra fields except through a validator, though.
(Warning, psudo code again... I don't entirely remember what pydantic calls and python magic attributes might exist): 
``` python

class AttributeSchemaWithExtra(AttributeSchema, extra=pydantic.Extra.allow):
    pass

@root_validator(pre=True)
def handle_extra_fields(cls, values):
    to_update = {}
    for field, data in values.items():
        if field not in KNOWN_FIELD_NAMES:
            normalized_data = AttributeSchemaWithExtra(**data)
            to_update[field] = normalized_data
    values.update(to_update)
    return values
```

So I think that's an outline of how we could solve this.  Hopefully my memory of hte things that pydantic can do is close enough to reality that something like this will work.

Note: Using a validator means that the generated json schema won't reflect these constraints so we should open a bug report to fix that in the future.

Reply:
> So, let's say there's a schema side (validating and normalizing the data) and a programming side (using the data after it's been normalized by the schema) to this problem.

The programming side depends on the questions in https://github.com/ansible-community/antsibull/pull/301#discussion_r691536718: how do we want to show all this data? Also we'd need some fallback mechanism (probably an empty doc fragment of the same name included in stable-2.9, stable-2.10 and stable-2.11) for this to be usable outside of ansible-core for a long time, if we want to avoid a lot of copy-pasta and potentially many different versions of the descriptions of some of the fields. I guess this is part of DaWGs (because it's documenation) and core (because this requires technical solutions for older ansible-core/-base/Ansible versions) territory and needs to be discussed with them.

Right now making sure that the existance of attributes doesn't prevent rendering of docs, i.e. the schema side, is probably the only part we can efficiently solve in the short term.

> The schema for that is probably a separate class that is the same as AttributeSchema but has `extra=Extra.allow`.
> [...]
> ```python
> class AttributeSchemaWithExtra(AttributeSchema, extra=pydantic.Extra.allow):
>     pass
> ```

Right now, the validation in ansible-core's validate-modules does not accept extra values for unknown attributes. I guess we'll have to decide whether we want to do that as well, or just allow everything for unknown attributes.